### PR TITLE
NNPACK bindings for Convolution

### PR DESCRIPTION
**Pull request by killeent**
This is going to change a lot.

cc @soumith

Will address #2810.

### CODE REVIEW

**Comment by Maratyszcza** [comment]
Two additional optimizations would help performance with NNPACK.
1. Pre-allocate workspace buffers. This is described in Maratyszcza/NNPACK#75
2. Pre-compute kernel transforms, in case when kernels do not change (i.e. inference, or "frozen" layers). See details in Maratyszcza/NNPACK#82

**Comment by killeent** [comment]
Some benchmark numbers from running 10 forward/backward passes on ResNet-50. Baseline is 75s, 313s for (1, 64) respectively.

| Threadpool Size | Re-use Workspace Buffers | batch size = 1 | batch size = 64 |
| ---------------- | ---------------------------| ---------------| ----------------|
| 4 | no | 72.09 | 317.51 |
| 16 | no | 54.79 | 248.55 |
| 4 | yes | 32.93 | 306.92 |
| 16 | yes | 21.77 | 207.45 |
| 24 | yes | 24.22 | 199. 1|

File: torch/csrc/autograd/functions/convolution.cpp [hunk 1]
```diff
@@ -807,13 +849,32 @@ static tensor_pair compute_grad_params(
...
+          nnpack::SpatialConvolution_accGradWeight(
+              input, grad_output, grad_weight,
+              kernel_size[1], kernel_size[0],
+              params.padding[1], params.padding[0]);
+
+          if (bias.defined() && params.should_compute_output(2)) {
+            // grad_output is in N, C, H, W, we re-shape and make contiguous
+            at::Tensor ones = grad_output.type().ones(input.size(0) * grad_output.size(2) * grad_output.size(3));
+            at::Tensor reshaped = grad_output.transpose(1, 3).contiguous().view({-1, ones.numel()});
+            grad_bias.addmv_(1.0, 1.0, reshaped, ones);
```
Comment (apaszke) [inline torch/csrc/autograd/functions/convolution.cpp]: is this really faster than a regular add with broadcasting?

File: torch/csrc/nnpack/NNPACK.cpp [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+namespace nnpack {
+
+// Stolen from Caffe2
+static pthreadpool_t nnpack_threadpool_ = nullptr;
+
+pthreadpool_t nnpack_threadpool() {
+  if (nnpack_threadpool_ == nullptr) {
+    enum nnp_status nnpack_status = nnp_initialize();
+    if (nnpack_status != nnp_status_success) throw std::runtime_error("could not initialize NNPack");
+    nnpack_threadpool_ = pthreadpool_create(16);
```
Comment (apaszke) [inline torch/csrc/nnpack/NNPACK.cpp]: I doubt this call cannot fail, so please add error checking

File: torch/csrc/nnpack/NNPACK.cpp [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+static inline void deallocate_workspace() {
+  if (workspace)
+    std::free(workspace);
+  workspace = nullptr;
+}
+
+static inline void allocate_workspace() {
+  if (workspace)
+    deallocate_workspace();
+  posix_memalign(&workspace, nnpack_memory_alignment_boundary, workspace_size);
```
Comment (apaszke) [inline torch/csrc/nnpack/NNPACK.cpp]: use `std::aligned_alloc`

File: torch/csrc/nnpack/NNPACK.cpp [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    throw std::runtime_error("NNPack convolutionOutput expects 4D input Tensor N,C,H,W");
+  }
+  // Our weight Tensor must be in the form oC,iC,kH,kW
+  if (weight.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D weight Tensor oC,iC,kH,kW");
+  }
+  // Our weight Tensor must be in the form N,oC,oH,oW
+  if (output.ndimension() != 4) {
+    throw std::runtime_error("NNPack convolutionOutput expects 4D output Tensor N,oC,oH,oW");
+  }
```
Comment (apaszke) [inline torch/csrc/nnpack/NNPACK.cpp]: Does nnpack have size checks? If not, we should check them ourselves

File: torch/csrc/nnpack/NNPACK.cpp [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+  if (input.type().ID() != at::TypeID::CPUFloat ||
+      weight.type().ID() != at::TypeID::CPUFloat ||
+      output.type().ID() != at::TypeID::CPUFloat ||
+      (bias.defined() && bias.type().ID() != at::TypeID::CPUFloat)) {
+    throw std::runtime_error("Mismatched Tensor types in NNPack convolutionOutput");
+  }
+
+  const size_t batch_size = input.sizes()[0];
+  const size_t input_channels = input.sizes()[1];
+  const size_t output_channels = weight.sizes()[0];
```
Comment (apaszke) [inline torch/csrc/nnpack/NNPACK.cpp]: Can't you just do `weight.size(0)`?

File: torch/csrc/nnpack/NNPACK.cpp [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+  if (workspace == nullptr) {
+    size_and_allocate_ws();
+  }
+
+  // Try to run with the newly created, or existing workspace
+  auto status = run();
+
+  if (status == nnp_status_insufficient_buffer) {
+    // Need to reallocate the workspace
+    deallocate_workspace();
```
Comment (apaszke) [inline torch/csrc/nnpack/NNPACK.cpp]: This should be done automatically in `allocate_workspace()` right?

File: torch/csrc/nnpack/NNPACK.cpp [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+    enum nnp_status nnpack_status = nnp_initialize();
+    if (nnpack_status != nnp_status_success) throw std::runtime_error("could not initialize NNPack");
+    nnpack_threadpool_ = pthreadpool_create(16);
+  }
+  return nnpack_threadpool_;
+}
+
+// Assuming for now this is Thread-safe, but that seems like a stretch
+static void *workspace = nullptr;
+static size_t workspace_size = 0;
```
Comment (apaszke) [inline torch/csrc/nnpack/NNPACK.cpp]: these two should either be `thread_local` or have a mutex guarding them

**Review by apaszke — COMMENTED** [review]
Looks good! Some minor comments

File: torch/csrc/nnpack/NNPACK.cpp (hunk 2 above)
Comment (killeent) [reply torch/csrc/nnpack/NNPACK.cpp]: Will do.

File: torch/csrc/nnpack/NNPACK.cpp (hunk 3 above)
Comment (killeent) [reply torch/csrc/nnpack/NNPACK.cpp]: I think `std::aligned_alloc` is C++17? http://en.cppreference.com/w/cpp/memory/c/aligned_alloc

File: torch/csrc/nnpack/NNPACK.cpp (hunk 4 above)
Comment (killeent) [reply torch/csrc/nnpack/NNPACK.cpp]: Nope, I can add more detailed checks now that we are cool with this overall approach.

File: torch/csrc/nnpack/NNPACK.cpp (hunk 5 above)
Comment (killeent) [reply torch/csrc/nnpack/NNPACK.cpp]: Will do.

**Comment by killeent** [comment]
If we are cool with this approach I think the only other thing to do is to add a threadpool # based on the host's properties, per @Maratyszcza's suggestion we could use his library https://github.com/Maratyszcza/cpuinfo but I was wondering if we had anything built into PyTorch already to detect the number of cores, etc.

File: torch/csrc/autograd/functions/convolution.cpp (hunk 1 above)
Comment (killeent) [reply torch/csrc/autograd/functions/convolution.cpp]: Do you mean making `ones` a Scalar?

**Comment by apaszke** [comment]
I think there are two simpler ways:
1. use `omp_get_num_threads()` - this is nice, because it can be set by the user if they want to restrict their job to fewer cores
2. [Use C++11 APIs for this](http://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency`)

**Comment by soumith** [comment]
using `omp_num_threads()` seems like the way to go.

File: torch/csrc/nnpack/NNPACK.cpp (hunk 3 above)
Comment (apaszke) [reply torch/csrc/nnpack/NNPACK.cpp]: Right, my bad. `posix_memalign` will unfortunately break Windows, so it would be nicer to use some standard function. Perhaps we could allocate `workspace_size + nnpack_memory_alignment_boundary` bytes and align them ourselves?

File: torch/csrc/autograd/functions/convolution.cpp (hunk 1 above)
Comment (apaszke) [reply torch/csrc/autograd/functions/convolution.cpp]: I mean `grad_bias = grad_output.sum(1)` or sth like this

File: torch/csrc/nnpack/NNPACK.cpp (hunk 6 above)
Comment (killeent) [reply torch/csrc/nnpack/NNPACK.cpp]: Yes, but we have to deallocate the workspace prior to the call otherwise the kernel will simply return insufficient buffer space again instead of calculating the new size that we need.

File: torch/csrc/nnpack/NNPACK.cpp (hunk 3 above)
Comment (killeent) [reply torch/csrc/nnpack/NNPACK.cpp]: I don't think NNPACK supports Windows either, so I don't think this will be a problem for now?

**Comment by fmassa** [comment]
nit: Do we want to add a flag to enable/disable NNPack at runtime, like we do for cudnn?
According to @alykhantejani experience using NNPack in Caffe2, for some configurations (apparently when the dimensions are not a power of two), NNPack can be significantly slower than the baseline convolution from Caffe2 with MKL.

**Comment by killeent** [comment]
@fmassa Yeah the other thing I wanted to say for us to do is add some sort of Python-level API for checking for availability (e.g. `torch.cuda.is_available()` equivalent) for unit testing. So we could potentially add that as well. I was just waiting until we had clarity on how we were going to make NNPACK available before doing so cc @soumith

File: torch/csrc/autograd/functions/convolution.cpp (hunk 1 above)
Comment (killeent) [reply torch/csrc/autograd/functions/convolution.cpp]: I will compare against transposing the channels to the innermost dim and doing a reduction as specified.

File: torch/csrc/nnpack/NNPACK.cpp (hunk 3 above)
Comment (fmassa) [reply torch/csrc/nnpack/NNPACK.cpp]: Can we explicitly disable NNPack on Windows then, to avoid compilation errors on Windows users? cc @peterjc123

**Comment by soumith** [comment]
@fmassa i thought NNPack is slower than MKLDNN, not MKL + Gemm

**Comment by fmassa** [comment]
@soumith yes, in Aly's tests NNPack was generally slower than MKLDNN. 
But if your input / weights have odd dimensions (for example, `1x3x63x63`), then NNPack seemed to be slower than unfold + MKL Gemm (and MKLDNN was even slower).
@alykhantejani can you give more details on that?

File: torch/csrc/autograd/functions/convolution.cpp (hunk 1 above)
Comment (killeent) [reply torch/csrc/autograd/functions/convolution.cpp]: In my test, the matrix multiply version was slightly faster or the same for a variety of different Conv params.

File: tools/setup_helpers/nnpack.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import os
+
+from .env import check_env_flag
+
+WITH_NNPACK = False
+NNPACK_LIB_DIR = None
+NNPACK_INCLUDE_DIRS = None
+
+if not check_env_flag('NO_NNPACK'):
```
Comment (peterjc123) [inline tools/setup_helpers/nnpack.py]: Please explicitly add `and` conditions for non-windows check here.

**Comment by Maratyszcza** [comment]
Some random notes:
1. `nnp_convolution_output`, `nnp_convolution_input_gradient`, `nnp_convolution_kernel_gradient` in NNPACK are not efficient when batch size, # input channels, or # output channels is small. Thus, you may want to add an option to enable/disable NNPACK on per-layer basis. You may also let users specify NNPACK algorithm on per-layer basis, default settings is not always the best.
2. For `nnp_convolution_output` with batch size 1, `nnp_convolution_inference` nearly always would result in better performance. NNPACK doesn't, however, provide similar functions for backward pass with batch size 1.
3. Do you use arbitrary tensor strides? I could make them work in NNPACK.

**Comment by soumith** [comment]
@Maratyszcza 
on (1), 
it's a terrible idea to let users specify NNPACK algorithm on a per-layer basis. Extending this concept, we'll eventually get to: something like [the TensorFlow performance guide](https://www.tensorflow.org/performance/performance_guide).
We need to come up with simple heuristics that predict at runtime if NNPACK is going to be faster. This shouldn't be that hard, as long as we take into account cache sizes, number of cores etc.

3. yep users can give input/output tensors of arbitrary strides.

File: torch/csrc/autograd/functions/convolution.cpp (hunk 1 above)
Comment (apaszke) [reply torch/csrc/autograd/functions/convolution.cpp]: Did you include the cost of allocating `ones` in here?

Comment (killeent) [reply torch/csrc/autograd/functions/convolution.cpp]: Yes - it was testing the entire forward and backward passes, not just the particular function call.

**Comment by alykhantejani** [comment]
@soumith yup, in some settings NNPACK is slower than im2col + gemm i.e. a conv with input `[1,3, 64, 64]` through a conv `inChannels=3, outChannels=64, kernelSize=3` I get the following times (averaged over 100 runs):

```
[DEFAULT]Avg iteration speed=0.3841ms
[NNPACK] Avg iteration speed=2.1265ms 	 speedup factor=0.180653
[MKLDNN] Avg iteration speed=0.3719ms 	 speedup factor=1.032965
```

where `DEFAULT` is caffe2's im2col + gemm. As you can see NNPack is much slower in this case

**Comment by killeent** [comment]
Per discussion with @soumith we don't have plans to integrate MKL-DNN due to the unfriendliness of the API.

**Comment by apaszke** [comment]
Also 3% speedup is really minor. We should only make sure that our CPU-side conv has similar perf to Caffe

**Comment by soumith** [comment]
@killeent you can repurpose this script by @alykhantejani to figure out decent heuristics for enabling NNPACK before committing it in: https://gist.github.com/alykhantejani/334ea077346288a93cbac142f50e92f4

**Comment by killeent** [comment]
Update: I used @alykhantejani's script for some benchmarking and found that `batch_size=32` is a good starting point. I want to bind the NNPACK `nnp_convolution_inference` to see how performance compares for `batch_size=1`. Then I will pick a final threshold for things and we can get this merged :)

**Comment by killeent** [comment]
Okay, I have finished benchmarking and have set the threshold for the `batch_size` to be 16, to enable NNPACK. I also bound the inference NNPACK call, although it will not be enabled for now as it relies upon `batch_size=1` - but I wanted to check and see what the difference was.

I think this is ready to go, cc @soumith

**Comment by fmassa** [comment]
Can we enable NNPack with `inference` for batch size 1? I think it would be a valuable addition, as the use-cases of running convolutions on the CPU usually involve a batch size of 1.

**Comment by killeent** [comment]
@fmassa - its not consistently outperforming our existing CPU kernel for some of the standard Vision models. For example, our existing kernel outperforms NNPACK on `resnet-18`, `alexnet`, and is a toss-up for `densenet` and some others. It could possibly be because @Maratyszcza has optimized for running inference, and thus does not have `batch_size=1` optimized kernel for the gradients.

**Comment by fmassa** [comment]
But batch size 1 for training doesn't work due to batch norm, right?
If NNPack is still slower for forward-only, then I'm ok just defaulting to our implementation.
